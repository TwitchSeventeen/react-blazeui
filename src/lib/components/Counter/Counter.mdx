---
name: Counter
route: /components/counter
menu: Components
---

import { Playground, PropsTable } from 'docz'
import Counter from './index';

# Counter

When you want to display a number with a bit more flair than usual.

## Basic usage
<Playground>
    <Counter start={120} value={0} duration={10} decimals={2} delay={2} prefix="Â£" suffix=" in bank account" />
</Playground>


## Properties
Under the hood this component uses [React CountUp](https://github.com/glennreyes/react-countup). 
Any of its additional [props](https://github.com/glennreyes/react-countup#props) can be added and will be passed down the component.
<PropsTable of={Counter} />