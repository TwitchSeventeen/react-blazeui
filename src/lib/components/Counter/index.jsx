import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import CountUp from 'react-countup';

function Counter(props) {
  const {
    className,
    start,
    value,
    duration,
    delay,
    prefix,
    as: Component,
    dataName,
    suffix,
    onStart,
    onEnd,
    ...other
  } = props;
  const classes = classNames(
    'c-counter',
    'u-super',
    'u-text--mono',
    className,
  );

  return (
    <Component className={classes} data-name={dataName}>

      { prefix
            && <span className="c-counter__prefix">{prefix}</span>}

      <CountUp
        {...other}
        className="c-counter__amount"
        start={start}
        end={value}
        duration={duration}
        delay={delay}
        onStart={onStart}
        onEnd={onEnd}
      />

      { suffix
        && <span className="c-counter__suffix u-text--quiet u-xsmall">{suffix}</span>}

    </Component>
  );
}

Counter.propTypes = {
  /**
   * Additional classes
   */
  className: PropTypes.string,
  /**
   * The initial value of the counter
   */
  start: PropTypes.number,
  /**
   * The ending value of the counter
   */
  value: PropTypes.number,
  /**
   * The time in seconds the transition runs
   */
  duration: PropTypes.number,
  /**
   * The delay in seconds before starting
   */
  delay: PropTypes.number,
  /**
   * Text to go before the value
   */
  prefix: PropTypes.string,
  /**
   * Text to go after the value
   */
  suffix: PropTypes.string,
  /**
   * Function called on transition start
   */
  onStart: PropTypes.func,
  /**
   * Function called on transition end
   */
  onEnd: PropTypes.func,
  /**
   * Optional data-name attribute
   * (use for testing/selection)
   */
  dataName: PropTypes.string,
  /**
   * Element type to render as
   */
  as: PropTypes.elementType,
};

Counter.defaultProps = {
  className: null,
  start: 0,
  value: 0,
  duration: null,
  delay: null,
  prefix: null,
  suffix: null,
  onStart: undefined,
  onEnd: undefined,
  dataName: 'counter',
  as: 'div',
};

export default Counter;
