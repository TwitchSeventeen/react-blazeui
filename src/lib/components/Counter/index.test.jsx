import React from 'react';
import { render, cleanup } from '@testing-library/react';
import Counter from './index';

describe('Counter', () => {
  afterEach(cleanup);

  it('should be defined', () => {
    expect(Counter).toBeDefined();
  });

  it('renders correctly', () => {
    const { asFragment } = render(
      <Counter start={100} value={0} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly with prefix and suffix', () => {
    const { asFragment } = render(
      <Counter start={100} value={0} prefix="Â£" suffix=" lorem" />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
