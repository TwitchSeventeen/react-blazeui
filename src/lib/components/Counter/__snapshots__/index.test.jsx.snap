// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter renders correctly 1`] = `
<div
  className="c-counter u-super u-text--mono"
>
  <CountUp
    className="c-counter__amount"
    decimal="."
    decimals={0}
    delay={null}
    duration={null}
    easingFn={null}
    end={0}
    formattingFn={null}
    onEnd={null}
    onPauseResume={[Function]}
    onReset={[Function]}
    onStart={null}
    onUpdate={[Function]}
    prefix=""
    redraw={false}
    separator=""
    start={100}
    suffix=""
    useEasing={true}
  />
</div>
`;

exports[`Counter renders correctly with prefix and suffix 1`] = `
<div
  className="c-counter u-super u-text--mono"
>
  <span
    className="c-counter__prefix"
  >
    Â£
  </span>
  <CountUp
    className="c-counter__amount"
    decimal="."
    decimals={0}
    delay={null}
    duration={null}
    easingFn={null}
    end={0}
    formattingFn={null}
    onEnd={null}
    onPauseResume={[Function]}
    onReset={[Function]}
    onStart={null}
    onUpdate={[Function]}
    prefix=""
    redraw={false}
    separator=""
    start={100}
    suffix=""
    useEasing={true}
  />
  <span
    className="c-counter__suffix u-text--quiet u-xsmall"
  >
     lorem
  </span>
</div>
`;
